torch/lib/THD/base/data_channels/DataChannelNccl.cpp

```diff
@@ -0,0 +1,715 @@
+  if (groupId != THDGroupWORLD &&
+      _groups.find(groupId) != _groups.end()) {
+    _groups.erase(groupId);
+  }
+  if (_groupDevices.find(groupId) != _groupDevices.end()) {
+    _groupDevices.erase(groupId);
+  }
+}
+
+
+// Initialization function
+bool DataChannelNccl::init() {
+
+  std::vector<rank_type> ranks;
+  ranks.reserve(_numProcesses);
+
+  for (rank_type rank = 0; rank < _numProcesses; ++rank) {
+    ranks.push_back(rank);
+  }
+
+  // Insert the current group
+  _groups.insert({
+    THDGroupWORLD,
+    DataChannel::Group(ranks, _numProcesses - 1)
+  });
+
+  // Get the GPU count
+  THCudaCheck(cudaGetDeviceCount(&_numGPUs));
+
+  return true;
+}
+
+
+rank_type DataChannelNccl::getRank() {
+  return _rank;
+}
+
+
+rank_type DataChannelNccl::getNumProcesses() {
+  return _numProcesses;
+}
+
+
+NcclResourcePair DataChannelNccl::_getNcclResourcePair(
+    std::vector<at::Tensor>& input,
+    THDGroup groupId) {
+
+  if (input.empty()) {
+    throw std::runtime_error("Not able to create/get the Nccl Comm since "
+                             "input tensor is empty");
+  }
+  // Get the deviceList String
+  std::string deviceList;
+  for (auto tensor : input) {
+    if (deviceList.empty()) {
+      deviceList = std::to_string(tensor.get_device());
+    } else {
+      deviceList += "," + std::to_string(tensor.get_device());
+    }
+  }
+
+  if (_groupDevices.find(groupId) != _groupDevices.end() &&
+      deviceList != _groupDevices[groupId]) {
+    std::string errMsg;
+    errMsg = "The current group: " + std::to_string(groupId) +
+            " has already got a GPU device list: " + _groupDevices[groupId]
+            + " associated with. Each group should only be associated with a "
+            + "given device list. Please create a new group for your provided "
+            + "device list: " + deviceList;
+    throw std::runtime_error(errMsg);
+  }
+
+  if (_groupNcclResources.find(groupId) != _groupNcclResources.end()) {
+    return std::make_pair(_groupNcclResources[groupId].ncclComms(),
+                          _groupNcclResources[groupId].ncclCudaEvents());
+  }
+
+  // Add in the device list of the group
+  _groupDevices[groupId] = deviceList;
+
+  // NCCL communicator
+  auto comms =
+    std::unique_ptr<std::vector<ncclComm_t>>(new std::vector<ncclComm_t>());
+
+  comms->resize(input.size());
+
+  // Corresponding CUDA events
+  auto events =
+    std::unique_ptr<std::vector<cudaEvent_t>>(new std::vector<cudaEvent_t>());
+
+  events->resize(input.size());
+
+  // Create the unique NCCL ID and broadcast it
+  ncclUniqueId ncclId;
+  NCCL_CHECK(ncclGetUniqueId(&ncclId));
+
+  // Broadcast so that each process can have a unique NCCL ID
+  broadcastUniqueNcclId(&ncclId);
+
+  // Guard GPU device
+  AutoGPU gpuGuard;
+
+  // Now creating the CUDA events
+  for (size_t i = 0; i < input.size(); ++i) {
+    gpuGuard.setDevice(input[i].get_device());
+    THCudaCheck(cudaEventCreate(&((*events)[i])));
```

Comment:
`&((*events)[i])` is exactly why I'd prefer to keep vectors as plain members :wink: